### ISSUE TITLE
Overrides of Unbound Config

### ISSUE THREAD

**Issue by consultantRR** [MatthewVance/unbound-docker#1]
HI there,

Is there a clean best practices approach to overriding various elements unbound config that you are specifying in `unbound, sh` at runtime without having to replace your entire  `/opt/unbound/etc/unbound/unbound.conf` file?

**Comment by MatthewVance** [comment]
No, I have not implemented anything like that though it’s a good idea. Feel free to submit a pull request if you implement it.

**Comment by consultantRR** [comment]
WIll do, thanks.

**Comment by MatthewVance** [comment]
@ProbablyRusty in my comment on #2, I gave some guidance on how this could be done. That pull request still had users replacing the entire file which is less than the ideal way to let users customize the config.

**Comment by hanscees** [comment]
Hi,
whats wrong with firing up the container once and then rewriting the config file? 
It might be practical to copy in a couple of default config files for people to choose from?
For instance 
unbound.example.forwarder
unbound.example.resolver

and so on? 
I have construed a yaml file together with a config I am happy to share if you like.

**Comment by MatthewVance** [comment]
Hi @hanscees, 

I don't think anything is wrong with that, but I think it is outside the scope of what I want to maintain. Also, the current setup provides a default config for secure forwarding and a way to override the default config with a custom one so I'm not sure how much additional benefit this would add. 

I imagine most people who want to run their own resolver won't need example ones to choose from. Unbound is well documented and sites like http://calomel.org already provide examples.

If this provides a benefit that I'm missing, let me know.

I'm more intrigued with letting certain values be set via environmental variables at run time. Something like: https://github.com/obi12341/docker-unbound/blob/master/start.sh.

I haven't taken the time to explore that beyond seeing what other systems have done. I previously [mentioned](https://github.com/MatthewVance/unbound-docker/pull/2#issuecomment-428777905) Postgres as an example.

**Comment by hanscees** [comment]
its a free world. 

but yml somehow does this too. 

This is my yml file for nextcloud. It passes on variables to the container. Thats to the mariadb container. 
Somehow they parse this. This seems like a simple method to me. But I am not a programmer.



`version: '2'

services:
  db:
    image: mariadb
    restart: always
    container_name: db$Today
    volumes:
      - nextcloud_db_data:/var/lib/mysql
    environment:
      - MYSQL_ROOT_PASSWORD=Vdamnsecret
      - MYSQL_PASSWORD=moredamnsecret 
      - MYSQL_DATABASE=nextcloud
      - MYSQL_USER=nextcloud
    networks:
      - isolated_nw

  app:
    image: nextcloud
    restart: always
    container_name: nextcloud$Today
    ports:
      - 8080:80
    links:
      - db
    volumes:
      - nextcloud_data:/var/www/html/
      - nextcloud_conf:/etc/apache2/
      - nextcloud_log:/var/log/
    restart: always
    networks:
      - isolated_nw

volumes:
  nextcloud_db_data:
  nextcloud_data:
  nextcloud_conf:
  nextcloud_log:
`

**Comment by zilexa** [comment]
As a compromise, may I suggest configuring the Docker image to simply store its default conf file in a folder that can be volume mapped by the user via a Volume variable?
For example (compose):
```
...
    container_name: unbound
    image: mvance/unbound:latest
    restart: always
    mac_address: d0:ca:ab:cd:ef:02
    networks:
      DNS-network:
        ipv4_address: 192.168.88.89
    volumes:
      - $USERDIR/docker/unbound:/opt/unbound/etc/unbound/
    ports:
      - 53/tcp
      - 53/udp
```

This way the default config is easily accessible from the host machine and can be overridden if necessary.

**Comment by MatthewVance** [comment]
Yes, I was thinking some rework was needed to better accommodate scenarios
like this one. The approach used the standalone Unbound setup should work
here.
https://github.com/MatthewVance/unbound-docker

**Comment by zilexa** [comment]
Exactly, however if I do that now, I have to create the conf files myself from scratch. I am only suggesting to store the default files in the volume mapped dir when the container starts. So that users can access/see/edit the configs easily.

**Comment by rochmind** [comment]
was going to use custom conf but then found forward-records.conf -- how to use?

**Comment by MatthewVance** [comment]
Make a copy of the file, edit to your liking, and mount as a volume.

docker run --name my-unbound -d -p 53:53/udp -p 53:53/tcp -v \
$(pwd)/forward-records.conf:/opt/unbound/etc/unbound/forward-records.conf:ro
\
--restart=always mvance/unbound:latest

https://github.com/MatthewVance/unbound-docker#override-default-forward

**Comment by rochmind** [comment]
ok thanks - does it mean if I use the default, it uses cloudfare as the upstream recursive server?

**Comment by MatthewVance** [comment]
That's right

**Comment by toastie89** [comment]
Hi! Would a `include: /opt/unbound/etc/unbound/*.conf` in the end of the configuration file make the trick? This way I could put what ever custom configuration in the directory in addition to the defaults?
